keps/sig-apps/3998-job-success-completion-policy/README.md

```diff
@@ -0,0 +1,681 @@
+	// JobSucceeded means the job has been succeeded.
+	JobSucceeded JobConditionType = "Succeeded"
+	...
+)
+```
+
+Moreover, we validate the following constraints for the `target`:
+- whether the specified indexes in the `succeededIndexes` and
+  the number of indexes in the `count` don't exceed the value of `completions`.
+- whether `Indexed` is specified in the `completionMode` field.
+- whether the size of `succeededIndexes` is under 64Ki.
+- exactly one of the enum `continue` or `terminate` is specified
+in the `whenTargetAchieved`.
+- whether the `succeededIndexes` field has a valid format.
+- whether the `count` field has an absolute number.
+- whether the target order hasn't changed.
+
+In addition, when the `successPolicy` isn't nil, we validate the `target` isn't nil.
+
+### Evaluation
+
+The job-controller evaluates the successPolicies every in updating pod conditions in order,
+and when the job meets one of successPolicies, other successPolicies are ignored.
+After that, we add `Succeeded` condition instead of `Failed` condition to `.status.conditions` even if some indexes fail.
+
+If a set of indexes are specified in the `indexes` field,
+the job-controller evaluates `.status.completedIndexes` to see if a set of indexes is there.
+
+If the number of indexes is specified in the `count`, 
+the job-controller evaluates `.status.succeeded` to see if the value is `count` or more.
+
+In addition, the job controller performs actions against running pods based on the value of `whenTargetAchieved`:
+
+- `continue`: Nothing. Running pods continues to run.
+- `terminate`: Not finished pods would be terminated by the job controller. Not finished pods means running or pending pods.
+
+#### The situations where successPolicy conflicts other terminating policies
+
+The successPolicy has potential conflicts with other terminating policies such as the [pod failure policy](https://kubernetes.io/docs/tasks/job/pod-failure-policy/)
+, backoffLimit, and [backoffLimitPerIndex](https://kubernetes.io/docs/concepts/workloads/controllers/job/#backoff-limit-per-index)
+in the following situations:
+
+- when the job meets the successPolicy and some pod failure policies with the `FailJob` action.
+- when the job meets the successPolicy and the number of failed pods exceeds `backoffLimit`.
+- when the job meets the successPolicy and the number of failed pods per indexes exceeds `backoffLimitPerIndex` in all indexes.
+
+To avoid the above conflicts, terminating policies are evaluated the first before successPolicies.
+
+In addition, we replace [`finishedCondition *batchv1.JobCondition`](https://github.com/kubernetes/kubernetes/blob/922fbbf5b70d461118dc03e4eb2eb2ddc8762481/pkg/controller/job/job_controller.go#L130)
+with `finishedConditions []batchv1.JobCondition`. Then, when the job meets both a successPolicy 
+and terminating policies in the same reconcile cycle, we add `Failed` and `Succeeded` conditions to the `finishedCondition` and all pods are deleted.  
```

Comment:
I don't think we should add both conditions ever.

It makes more sense to me that the Failed condition takes precedence.

Reply:
Actually, I discussed this offline with @mimowo and @deads2k.
Indeed, we agreed with disallowed `Succeeded=true`, `Failed=true`, and `Completed=false`.

Since I will update this section, you can check the details afterward.

Reply:
Done.